src/libponyrt/ds/hash.c

```diff
@@ -22,13 +22,18 @@ static void* search(hashmap_t* map, size_t* pos, void* key, size_t hash,
   size_t index = ((hash & mask) + p_length) & mask;
   void* elem;
   size_t elem_hash;
+  size_t ib_index;
+  size_t ib_offset;
 
   for(size_t i = 0; i <= mask; i++)
   {
+    ib_index = index/HASHMAP_BITMAP_TYPE_SIZE;
+    ib_offset = index%HASHMAP_BITMAP_TYPE_SIZE;
     elem = map->buckets[index].ptr;
     elem_hash = map->buckets[index].hash;
 
-    if(elem == NULL)
+    // if bucket is empty
+    if((map->item_bitmap[ib_index] & ((bitmap_t)1 << ib_offset)) == 0)
```

Comment:
Then, separate out the next two cases from the `if`. That is, since the first case does a `return`, get rid of the `else`. Before the next two cases, fetch `elem_hash`, but not `elem`. Separate the last case as well (getting rid of the `else`), and only fetch `elem` if `hash == elem_hash` (i.e. change the `&&` into a nested `if` where `elem` is fetched in between).

Theoretically, that will lower the average memory access count. But only benchmarking lets us know if it's worth it :)